src/StandaloneModuleGraph.zig

```diff
@@ -824,4 +885,175 @@ pub const StandaloneModuleGraph = struct {
         }
     }
+
+    /// Source map serialization in the bundler is specially designed to be
+    /// loaded in memory as is. Source contents are compressed with ZSTD to
+    /// reduce the file size, and mappings are stored as uncompressed VLQ.
+    pub const SerializedSourceMap = struct {
+        bytes: []const u8,
+
+        /// Following the header bytes:
+        /// - source_files_count number of StringPointer, file names
+        /// - source_files_count number of StringPointer, zstd compressed contents
+        /// - the mapping data, `map_vlq_length` bytes
+        /// - all the StringPointer contents
+        pub const Header = extern struct {
+            source_files_count: u32,
+            map_bytes_length: u32,
+        };
+
+        pub fn header(map: SerializedSourceMap) *align(1) const Header {
+            return @ptrCast(map.bytes.ptr);
+        }
+
+        pub fn mappingVLQ(map: SerializedSourceMap) []const u8 {
+            const head = map.header();
+            const start = @sizeOf(Header) + head.source_files_count * @sizeOf(StringPointer) * 2;
+            return map.bytes[start..][0..head.map_bytes_length];
+        }
+
+        pub fn sourceFileNames(map: SerializedSourceMap) []align(1) const StringPointer {
+            const head = map.header();
+            return @as([*]align(1) const StringPointer, @ptrCast(map.bytes[@sizeOf(Header)..]))[0..head.source_files_count];
+        }
+
+        fn compressedSourceFiles(map: SerializedSourceMap) []align(1) const StringPointer {
+            const head = map.header();
+            return @as([*]align(1) const StringPointer, @ptrCast(map.bytes[@sizeOf(Header)..]))[head.source_files_count..][0..head.source_files_count];
+        }
+
+        /// Once loaded, this map stores additional data for keeping track of source code.
+        pub const Loaded = struct {
+            map: SerializedSourceMap,
+
+            /// Only decompress source code once! Once a file is decompressed,
+            /// it is stored here
+            decompressed_files: []?[]u8,
+
+            pub fn sourceFileContents(this: Loaded, index: usize) []const u8 {
+                if (this.decompressed_files[index]) |decompressed|
+                    return decompressed;
+
+                const compressed_codes = this.map.compressedSourceFiles();
+                const compressed_file = compressed_codes[@intCast(index)].slice(this.map.bytes);
+                const size = bun.zstd.getDecompressedSize(compressed_file);
+
+                const bytes = bun.default_allocator.alloc(u8, size) catch bun.outOfMemory();
+                const result = bun.zstd.decompress(bytes, compressed_file);
+
+                if (result == .err) {
+                    bun.Output.warn("Source map decompression error: {s}", .{result.err});
+                    bun.default_allocator.free(bytes);
+                    this.decompressed_files[index] = "";
+                    return "";
```

Comment:
This should return null so that the parent is aware the source map contents was null instead of an empty string